07-folds-monoids/JoinList.hs

```diff
@@ -0,0 +1,244 @@
+indexJ _ _ = Nothing
+
+
+
+
+{-
+Implement the function
+dropJ :: (Sized b, Monoid b) =>
+Int -> JoinList b a -> JoinList b a
+The dropJ function drops the first n elements from a JoinList. This is analogous to the standard drop function on lists. Formally, dropJ should behave in such a way that
+jlToList (dropJ n jl) == drop n (jlToList jl)
+-}
+dropJ :: (Sized b, Monoid b) => Int -> JoinList b a -> JoinList b a
+dropJ n (Single m a)
+    | n >= 1    = Empty
+    | otherwise = Single m a
+dropJ n (Append _ jl1 jl2)
+    | n < jl1Size = dropJ n             jl1 +++ jl2
+    | otherwise   = dropJ (n - jl1Size) jl2
+      where
+        jl1Size :: Int
+        jl1Size = getSize . size . tag $ jl1
+dropJ _ _ = Empty
+
+
+
+
+{-
+Finally, implement the function takeJ :: (Sized b, Monoid b) =>
+Int -> JoinList b a -> JoinList b a
+The takeJ function returns the first n elements of a JoinList, dropping all other elements. Again, this function works similarly to the standard library take function; that is, it should be the case that
+jlToList (takeJ n jl) == take n (jlToList jl).
+Ensure that your function definitions use the size function from the Sized type class to make smart decisions about how to descend into the JoinList tree.
+-}
+takeJ :: (Sized b, Monoid b) => Int -> JoinList b a -> JoinList b a
+takeJ n (Single m a)
+    | n <= 0    = Empty
+    | otherwise = Single m a
+takeJ n (Append _ jl1 jl2)
+    | n < jl1Size = takeJ n jl1
+    | otherwise   = jl1 +++ takeJ (n - jl1Size) jl2
+      where
+        jl1Size :: Int
+        jl1Size = getSize . size . tag $ jl1
+takeJ _ _ = Empty
+
+--Exercise 3
+scoreLine :: String -> JoinList Score String
+scoreLine str = Single (scoreString str) str
+
+
+{-
+Exercise 4 Finally, combine these two kinds of annotations. A pair of monoids is itself a monoid:
+ instance (Monoid a, Monoid b) => Monoid (a,b) where
+   mempty = (mempty, mempty)
+   mappend (a1,b1) (a2,b2) = (mappend a1 a2, mappend b1 b2)
+(This instance is defined in Data.Monoid.) This means that join-lists can track more than one type of annotation at once, in parallel, sim- ply by using a pair type.
+Since we want to track both the size and score of a buffer, you should provide a Buffer instance for the type
+JoinList (Score, Size) String.
+Due to the use of the Sized type class, this type will continue to work with your functions such as indexJ.
+Finally, make a main function to run the editor interface using your join-list backend in place of the slow String backend (see StringBufEditor.hs for an example of how to do this). You should create an initial buffer of type JoinList (Score, Size) String and pass it as an argument to runEditor editor. Verify that the editor demonstration described in the section “Editors and Buffers” does not exhibit delays when showing the prompt.
+-}
+--Exercise 4
+instance Buffer (JoinList (Score, Size) String) where
+    toString = unlines . fromDList . jlToDList
+
+    fromString = foldl' func Empty . lines
```

Comment:
This is technically correct, but you end up with a `JoinList` that doesn't take advantage of the tree nature of the `JoinList` data structure, and is instead a linear structure (isomorphic to a list of the lines). Can you think of a better way to do this, such that you end up with a `JoinList` where the right and left `JoinList`s of the root node are approximately the same size, and this property holds true for each node?

Reply:
@oishikg
Something like this?
```haskell
insert :: String -> JoinList (Score, Size) String -> JoinList (Score, Size) String
insert str Empty = Single (scoreString str, 1) str
insert str jl@(Single _ _) = jl +++ Single (scoreString str, 1) str
insert str jl@(Append m jl1 jl2)
    | getJLSizeInt jl1 > getJLSizeInt jl2 = Append updatedMonoid jl1              (insert str jl2)
    | otherwise                           = Append updatedMonoid (insert str jl1) jl2
      where
        updatedMonoid :: (Score, Size)
        updatedMonoid = (scoreString str <> getJLScore jl, 1 <> getJLSize jl)

        getJLScore :: (Scored b, Monoid b) => JoinList b a -> Score
        getJLScore = scoreAttr . tag

        getJLSize :: (Sized b, Monoid b) => JoinList b a -> Size
        getJLSize = size . tag

fromString = foldr insert Empty . lines
```

Reply:
@peironggg, this solution would indeed adhere to the balancedness property that I stated above. However, if I'm not wrong, the following property should hold with regards to the `fromString` and `toString` functions: 

```haskell
-- For any string s, it should be the case that:
s = toString $ fromString s
```

Do you think your solution has this property? If not, can you think of a solution that has this property _and_ constructs a balanced `JoinList`? As a hint, consider the following problem: Given a list of some elements, how would you insert them into a binary tree such that:
- The tree is balanced 
- When the tree is flattened into a list, you get back the same list

The flatten function would be defined as so:

```haskell
flatten Leaf = []
flatten (Node l v r) = flatten l ++ [v] ++ flatten r
```

Reply:
@oishikg Thanks for taking the time to explain 👍 
This is what I came up with:

```haskell
buildBalanced :: [String] -> JoinList (Score, Size) String
buildBalanced []   = Empty
buildBalanced strs = buildBalanced left +++ Single (scoreString str, 1) str +++ buildBalanced right
  where
    (left, rightPart) = splitAt ((length strs `div` 2) - 1) strs

    right :: [String]
    right = tail rightPart

    str :: String
    str = head rightPart -- safe to use head here as splitting always leaves at least 1 in right half

    fromString = buildBalanced . lines
```
Testing in `ghci` appears to give the expected outcome:
```haskell
*JoinList> (toString . buildBalanced . lines) defaultText -- where fromString = buildBalanced . lines
"This buffer is for notes you don't want to save, and for\nevaluation of steam valve coefficients.\nTo load a different file, type the character L followed\nby the name of the file.\n"
*JoinList> defaultText
"This buffer is for notes you don't want to save, and for\nevaluation of steam valve coefficients.\nTo load a different file, type the character L followed\nby the name of the file.\n"
```